module/canvas/token-placement.mjs

```diff
@@ -0,0 +1,259 @@
+
+  /**
+   * Track the timestamp when the last mouse move event was captured.
+   * @type {number}
+   */
+  #moveTime = 0;
+
+  /* -------------------------------------------- */
+
+  /**
+   * Placements that have been generated.
+   * @type {PlacementData[]}
+   */
+  #placements;
+
+  /* -------------------------------------------- */
+
+  /**
+   * Preview tokens. Should match 1-to-1 with placements.
+   * @type {Token[]}
+   */
+  #previews;
+
+  /* -------------------------------------------- */
+
+  /**
+   * Determine whether the throttling period has passed. Set to `true` to reset period.
+   * @type {boolean}
+   */
+  get #throttle() {
+    return Date.now() - this.#moveTime <= 20;
+  }
+
+  set #throttle(value) {
+    this.#moveTime = Date.now();
+  }
+
+  /* -------------------------------------------- */
+  /*  Placement                                   */
+  /* -------------------------------------------- */
+
+  /**
+   * Perform the placement, asking player guidance when necessary.
+   * @param {TokenPlacementConfiguration} config
+   * @returns {PlacementData[]}
+   */
+  static place(config) {
+    const placement = new this(config);
+    return placement.place();
+  }
+
+  /**
+   * Perform the placement, asking player guidance when necessary.
+   * @returns {PlacementData[]}
+   */
+  async place() {
+    this.#createPreviews();
+    try {
+      return await this.#activatePreviewListeners();
+    } finally {
+      this.#destroyPreviews();
+    }
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Create token previews based on the prototype tokens in config.
+   */
+  #createPreviews() {
+    this.#placements = [];
+    this.#previews = [];
+    for ( const prototypeToken of this.config.tokens ) {
+      const tokenData = prototypeToken.toObject();
+      if ( tokenData.randomImg ) tokenData.texture.src = prototypeToken.actor.img;
+      const cls = getDocumentClass("Token");
+      const doc = new cls(tokenData, { parent: canvas.scene });
+      this.#placements.push({ x: 0, y: 0, rotation: 0 });
+      this.#previews.push(doc);
+      doc.object.draw();
+    }
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Clear any previews from the scene.
+   */
+  #destroyPreviews() {
+    this.#previews.forEach(p => p.object.destroy());
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Pixel offset to ensure snapping occurs in middle of grid space.
+   * @param {PrototypeToken} token  Token for which to calculate the adjustment.
+   * @returns {{x: number, y: number}}
+   */
+  #getSnapAdjustment(token) {
+    const size = canvas.dimensions.size;
+    switch ( canvas.grid.type ) {
+      case CONST.GRID_TYPES.SQUARE:
+        return {
+          x: token.width % 2 === 0 ? Math.round(size * 0.5) : 0,
+          y: token.height % 2 === 0 ? Math.round(size * 0.5) : 0
+        };
+      default:
+        return { x: 0, y: 0 };
+    }
+  }
```

Comment:
We should be able to use `canvas.grid.getSnappedPosition` (`getSnappedPoint` in v12) for this. I don't think we need to handle this ourselves.

Reply:
I find that using `getSnappedPosition` works fine for 1x1 tokens, but results in larger or stranger shaped tokens being offset in rather than centered under the token. I tried this:

```javascript
const center = canvas.grid.getSnappedPosition(
  point.x - Math.round(canvas.dimensions.size * 0.5),
  point.y - Math.round(canvas.dimensions.size * 0.5),
  1,
  { token: this.config.tokens[0].object }
);
```

For 2x2 tokens for example, this would result in the token being places with the top-left quarter centered over the square that I am hovering which makes it non-intuitive where to move the mouse to place the token.

Reply:
Ah, that's fair enough. I think this has all been improved in v12 so I will experiment with checking `game.release` here and calling `getSnappedPoint` instead, but this seems good enough for now.